### DISCUSSION TITLE
Question about simplification

### DISCUSSION THREAD

**Discussion by proteusGIT** [cvc5/cvc5#7356] [Q&A] [answered]
Hi,

is there some mechanism in CVC5 to simplify expressions? Surely this is not the main purpose of CVC5 but I am wondering if there is any kind of support (available to me as a user) in that direction. I have terms with only real valued variables, linear arithmetic, and quantification.

Best

**Comment by martin-cs** [comment]
There are a number of things that might fit your needs.  cvc5 has a rewriter which does context-free and local rewrites.  So for example x + -x  would be rewritten to 0 but if you had formulae saying y = x and z = -x then it could not rewrite y + z to 0.  I think it is available directly as part of the API.

More sophisticated simplifications require keeping some context or combining formulae in non-local ways.  For example, there is an equality engine which tracks and substitutes equalities (and so could solve the previous example).  There are also a number of techniques used in the theory solvers.  As these keep and require an amount of context, it is harder to separate them from the rest of the solver.

Could you give some examples of the kinds of thing you would like to be able to do?

**Comment by proteusGIT** [comment]
I suppose that I would need a little more elaborate simplification.
I have attached an example of one of my terms in `term.txt`.
Isabelle can simplify this term (see the attached Isabelle file [change the ending to thy of course] to the following.
```
∃x2_2≤10. 0 ≤ x2_2 ∧ cts_0_13 ≤ x2_2 ∧ cts_0_4 ≤ x2_2 ∧ cts_0_11 ≤ x2_2 ∧ cts_0_12 ≤ x2_2
```

So I am wondering whether CVC5 can also simplify the term similarly. I am definitely willing to consider any trade of suboptimal simplification for speed and tool integration here.
However, ideally, I would like to not need to tweak the simplification process for every new example by (de)selecting tactics based on what happens (e.g. due to some tactics not terminating [as fast as expected]).
I would assume that a solver could determine eligible tactics from the datatypes/used operations/quantification already.

[term.txt](https://github.com/cvc5/cvc5/files/7328201/term.txt)
[IsabelleScratch.txt](https://github.com/cvc5/cvc5/files/7328212/IsabelleScratch.txt)

**Comment by ajreynol** [comment, marked as answer]
We do have an interface for simplification in our API (https://github.com/cvc5/cvc5/blob/master/src/api/cpp/cvc5.h#L3671) and the *.smt2 parser, here is an example:
```
(set-logic UF)
(declare-fun P (Int) Bool)
(simplify (forall ((x Int)) (=> (= x 0) (P x))))
```
Given this input, cvc5 outputs:
```
(P 0)
```
To make this simplification configurable, it would require our rewriter to be configurable via user-provided options.  Our rewriter is already to some extent configurable. For example, using `--no-var-elim-quant` on the example above, you'd get output:
```
(forall ((x Int)) (or (not (= x 0)) (P x)))
```
There are some options to control how quantifiers are simplified; arithmetic terms tend to be simplified in a straightforward way.
If there is any specific rewrites we don't incorporate that you are interested in, do let us know.

**Comment by proteusGIT** [comment]
I applied CVC4 (since there are no CVC5 binaries available for windows apparently and I do not have the time right now to somehow get the building working under windows) to the following file replacing XXX by the term in the file I posted above.
```
(set-logic LIRA)
(declare-fun cts-0-11 () Real)
(declare-fun cts-0-4 () Real)
(declare-fun cts-0-13 () Real)
(declare-fun cts-0-12 () Real)
(simplify XXX)
```
While the CVC output below is not human readable (which is not really a problem in my case but is there an option to disable the use of let in outputs?), compared to the output of Isabelle simplification above (`∃x2_2≤10. 0 ≤ x2_2 ∧ cts_0_13 ≤ x2_2 ∧ cts_0_4 ≤ x2_2 ∧ cts_0_11 ≤ x2_2 ∧ cts_0_12 ≤ x2_2`), the result of simplification is a bit underwhelming maybe. I can't even tell whether anything has been simplified or just compressed using more usage of let.
```
(forall ((x2-0 Real)) (let ((_let_0 (>= x2-0 0.0))) (let ((_let_1 (= x2-0 0.0))) (let ((_let_2 (not _let_1))) 
(let ((_let_3 (not _let_0))) (or _let_3 (and (or _let_2 _let_3) _let_2) (and _let_1 _let_0 (not (forall
 ((BOUND_VARIABLE_780 Real)) (let ((_let_4 (* (- 1.0) x2-0))) (let ((_let_5 (* (- 1.0) 
BOUND_VARIABLE_780))) (let ((_let_6 (>= BOUND_VARIABLE_780 0.0))) (let ((_let_7 (not _let_6))) 
(let ((_let_8 (+ _let_4 BOUND_VARIABLE_780))) (let ((_let_9 (>= _let_8 0.0))) (let ((_let_10
 (not _let_9))) (let ((_let_11 (+ x2-0 _let_5)))  (let ((_let_12 (>= _let_11 (- 10.0)))) (let ((_let_13 
(not _let_12))) (or _let_10 _let_13 (and _let_1 _let_0 _let_9 _let_12 (or (and _let_1 _let_6 _let_0 
_let_9 _let_12 (or _let_2 _let_3 _let_13 _let_7 _let_10)) (and (or _let_2 _let_7 _let_13 _let_3 _let_10 
(not (>= (+ (* (- 1.0) cts-0-11) BOUND_VARIABLE_780) 0.0))) _let_1 _let_6 _let_9 _let_0 _let_12) 
(and (or _let_2 _let_7 _let_13 _let_3 _let_10 (not (>= (+ (* (- 1.0) cts-0-4) BOUND_VARIABLE_780)
 0.0))) _let_1 _let_6 _let_9 _let_0 _let_12) (and (or _let_2 _let_7 _let_13 _let_3 _let_10 (not (>= (+
 (* (- 1.0) cts-0-13) BOUND_VARIABLE_780) 0.0))) _let_1 _let_6 _let_9 _let_0 _let_12) (and (or
 _let_2 _let_7 _let_13 _let_3 _let_10 (not (>= (+ (* (- 1.0) cts-0-12) BOUND_VARIABLE_780) 0.0)))
 _let_1 _let_6 _let_9 _let_0 _let_12))) (and _let_12 _let_0 _let_1 (not (forall ((x1-2 Real)) (let ((_let_2
 (+ x1-2 _let_5))) (let ((_let_3 (>= _let_2 0.0))) (let ((_let_6 (+ _let_4 x1-2))) (let ((_let_7 (>=
 _let_6 0.0))) (or (not _let_7) (and _let_0 _let_7 _let_12 (not _let_3) _let_1) _let_3))))) )))))))))))))) )))))))) )
```
Can you maybe check if CVC5 obtains better results?
Also, would LIRA be the optimal choice here?

**Comment by ajreynol** [comment]
The logic won't impact which rewrites are applied.  You can also try the option `--ext-rewrite-quant`.  This should make cvc5 apply more aggressive simplification rules to the bodies of quantified formulas.
